cli/src/program.rs

```diff
@@ -1129,22 +1289,173 @@ fn process_program_deploy(
+        // Using try_partial_sign here because fee_payer_signer might not be the fee payer we
+        // end up using for this transaction (it might be NullSigner).
+        tx.try_partial_sign(signers, blockhash)?;
+        return_signers_with_config(
+            &tx,
+            &config.output_format,
+            &ReturnSignersConfig {
+                dump_transaction_message: true, // print out resulting signature
+            },
+        )
+    } else {
+        let final_message = Some(final_message);
+
+        let buffer_len = {
+            // Check supplied buffer account.
+            if let Some(account) = rpc_client
+                .get_account_with_commitment(&buffer_pubkey, config.commitment)?
+                .value
+            {
+                if !bpf_loader_upgradeable::check_id(&account.owner) {
+                    return Err(format!(
+                        "Buffer account {buffer_pubkey} is not owned by the BPF Upgradeable Loader",
+                    )
+                    .into());
+                }
+
+                match account.state() {
+                    Ok(UpgradeableLoaderState::Buffer { .. }) => {
+                        // continue if buffer is initialized
+                    }
+                    Ok(UpgradeableLoaderState::Program { .. }) => {
+                        return Err(format!(
+                            "Cannot use program account {buffer_pubkey} as buffer"
+                        )
+                        .into());
+                    }
+                    Ok(UpgradeableLoaderState::ProgramData { .. }) => {
+                        return Err(format!(
+                            "Cannot use program data account {buffer_pubkey} as buffer",
+                        )
+                        .into())
+                    }
+                    Ok(UpgradeableLoaderState::Uninitialized) => {
+                        return Err(
+                            format!("Buffer account {buffer_pubkey} is not initialized").into()
+                        );
+                    }
+                    Err(_) => {
+                        return Err(format!(
+                            "Buffer account {buffer_pubkey} could not be deserialized"
+                        )
+                        .into())
+                    }
+                };
+
+                account
+                    .data
+                    .len()
+                    .saturating_sub(UpgradeableLoaderState::size_of_buffer_metadata())
+            } else {
+                return Err(format!(
+                    "Buffer account {buffer_pubkey} not found, was it already consumed?",
+                )
+                .into());
+            }
+        };
+
+        let min_rent_exempt_program_data_balance = rpc_client
+            .get_minimum_balance_for_rent_exemption(UpgradeableLoaderState::size_of_programdata(
+                buffer_len,
+            ))?;
```

Comment:
Hmm, I thought for some reason that "buffer data" gets copied into "program account", while what seems to be actually happening is "program account" is just referencing "program data account" (written during buffer create action),

anyway, changed inhttps://github.com/solana-labs/solana/pull/33860/commits/40c967526981a89397771d8857c03154eaf18193 (in accordance with your comment from below https://github.com/solana-labs/solana/pull/33860#discussion_r1412541088)

Reply:
The buffer data does get copied into the program data account, so you've understood that behavior. The difference is that this is checking that the payer has enough lamports to afford the rent-exemption for an account with `buffer_len` bytes.

Since the buffer has already been initialized with `write-buffer`, the payer does not need to pay for this account again, so this check is unnecessary, and will also give incorrect errors.